### PR TITLE
Feature/webserver returns algo results as json

### PR DESCRIPTION
**Pull request by pk-scs**
# Additional HTTP POST Handler (JSON Response)

Added additional HTTP POST handler to Tornado webserver that sends algorithm results as JSON in HTTP response.

- Gets algo input from HTTP POST request body (JSON)
- Returns 'report_data' in HTTP response body (JSON)

Note that for using this additional HTTP POST, either XSRF cookies must be turned off when starting Tornado or XSRF cookie must be obtained and sent in HTTP POST request header. XSRF cookie can be obtained by sending an HTTP GET request to e.g. '/calculator'.

Also added test that checks for presence of key 'prob_inf' in HTTP POST response body.

### CODE REVIEW

File: cara/apps/calculator/__init__.py
Comment (pelson) [inline cara/apps/calculator/__init__.py:151]: Is your objective really to allow the same **form** inputs to generate a report as well as JSON? (as implemented here)

I wonder if what you really want is a raw Model type (not ``FormData``, but an actual serialised ``cara.monte_carlo.models.ExposureModel`` instance)?

File: cara/tests/apps/calculator/test_report_json.py
Comment (pelson) [inline cara/tests/apps/calculator/test_report_json.py]: Some more validation of what the response type looks like would be worthy here.

File: cara/tests/apps/calculator/test_report_json.py
Comment (pelson) [inline cara/tests/apps/calculator/test_report_json.py]: Can you re-use the body from the ``baseline_form_data`` fixture, rather than having an explicit duplication of the input here?

File: cara/apps/calculator/__init__.py
Comment (pelson) [inline cara/apps/calculator/__init__.py:263]: Normally I would suggest the return type doesn't belong in the URL, rather it is a request parameter (``content-type``). Is there a good reason for making it this way?

Comment (pelson) [reply cara/apps/calculator/__init__.py:263]: One reason, may be: we should be using the GET endpoint, with input JSON as an input argument. This would mean we also don't need to be able to toggle the XSRF cookies as well.

File: cara/apps/calculator/__init__.py
Comment (pelson) [inline cara/apps/calculator/__init__.py]: See my other comment about the endpoint being exposed (via GET), I don't believe this is the right thing to be doing.

**Comment by pelson** [comment]
Looks good! I think there are a few fundamental questions to answer before we hit on the best solution all around, but that should be fairly small realignments, rather than major refactors! Nice work :+1: 

I'm really interested to see what would be involved in the HTML report making use of this JSON endpoint also. It seems that would be a nice way to break the processing problem of the HTML report down into component parts. Not for this PR, but interesting to think about :+1:

File: cara/apps/calculator/__init__.py
Comment (pk-scs) [reply cara/apps/calculator/__init__.py:151]: The intention, from the discussions with ARVE, is indeed to allow for exactly the same inputs that are specified through the HTML form. 

From a conceptual point of view, indeed, the detour over `FormData` might not be needed. In the end, the JSON endpoint "builds" a `cara.monte_carlo.models.ExposureModel` from the `FormData`. However, converting the HTTP request body into a `cara.monte_carlo.models.ExposureModel` instance would duplicated existing logic on `FormData`. This logic could be extracted into an own "entity" and re-used. However, in the end the data in the request body represents form data. I guess it could be OK to convert it into a `FormData` instance first. Or maybe rename `FormData` into `ModelData`?

File: cara/tests/apps/calculator/test_report_json.py
Comment (pk-scs) [reply cara/tests/apps/calculator/test_report_json.py]: I see the point. In the end, at least for ARVE, the "entity" that will be further processed is the infection probability (`prob_inf`). I extended the validation now to:  

```
data = json.loads(response.body)
self.assertIsInstance(data['prob_inf'], float)
self.assertIsInstance(data['expected_new_cases'], float)
```
Does it make sense from your point of view to validate more of the response? If yes, let me know, then I'll add it to the test.

File: cara/tests/apps/calculator/test_report_json.py
Comment (pk-scs) [reply cara/tests/apps/calculator/test_report_json.py]: Done.

File: cara/apps/calculator/__init__.py
Comment (pk-scs) [reply cara/apps/calculator/__init__.py:263]: I changed the endpoint now from POST to GET. Indeed, this solves the issues with the XSRF cookie. As I understand, it's not standard practice to send a request body with a HTTP GET method. But it's not "forbidden" either. In our case, defining an additional GET endpoint might be the better solution.

File: cara/apps/calculator/__init__.py
Comment (pk-scs) [reply cara/apps/calculator/__init__.py]: See my comment above. Thank you for the code review 👍

File: cara/apps/calculator/__init__.py
Comment (pelson) [reply cara/apps/calculator/__init__.py:151]: I have no objection to using ``FormData`` here, just raising the question as to what the most useful thing is for ARVE. Keep this in mind in the future if you start finding that ``FormData`` is too limited compared to what a direct model could actually deliver.

For my own curiosity, how do you intend to build the form payload for your application? Are you going to generate it once and then programmatically inject the parameters you care about? I ask because this increases the importance of ``FormData`` stability (which is something we have taken care of in the past, but will need to be extra careful of in the future)

File: cara/apps/calculator/__init__.py
Comment (pelson) [reply cara/apps/calculator/__init__.py:263]: > it's not standard practice to send a request body with a HTTP GET method

Agreed (https://stackoverflow.com/a/983458/741316). I actually meant passing it as a query string to a GET request. This however has its own downsides (query string limit in practice is ~1024 chars), which may mean that we can't do that. In order of preference:

```
GET + querystring > POST + body + xsrf disabled > GET + body
```

If we need to go back to POST then we can disable XSRF on a per-endpoint basis by implementing the ``ConcentrationModelJsonResponse.check_xsrf_cookie`` method (https://github.com/tornadoweb/tornado/issues/254#issuecomment-147694082).

File: cara/tests/apps/calculator/test_report_json.py
Comment (pelson) [reply cara/tests/apps/calculator/test_report_json.py]: I think this is fine for now. We may want to increase it in the future, but no need just yet.

**Comment by pk-scs** [comment]
Thanks for the review and apologies for my late reply! Me and my family have been sick for the last several days. According to the tests we did it was not Covid-19. I guess the "usual" flew-like diseases didn't disappear just because of Covid-19 😅

File: cara/apps/calculator/__init__.py
Comment (pk-scs) [reply cara/apps/calculator/__init__.py:151]: The intention is to generate the form payload from within the ARVE application automatically. The application runs on an embedded device that is placed in a room (e.g. hotel or conference room). The device contains several sensors, which measure air quality etc. The idea is then to generate the form payload from 

- the "configuration" of the room the device is place in
- sensor data of the device

File: cara/apps/calculator/__init__.py
Comment (pk-scs) [reply cara/apps/calculator/__init__.py:263]: I think I would prefer `POST + body + xsrf disabled` or `GET + body` over `GET + querystring`. For the same reasons you mention in your comment (querry string limit). In that case, I think the `POST + body + xsrf disabled` would be the preferred solution. Should I go back to the POST then trying to disabling XSRF on a per-endpoint basis as you suggested?

Comment (pelson) [reply cara/apps/calculator/__init__.py:263]: Yes please, this would be best (POST + body + per-endpoint XSRF disable). Apologies for the flip-flop.

File: cara/apps/calculator/__init__.py
Comment (pelson) [reply cara/apps/calculator/__init__.py:151]: :+1: - thanks for the context.

File: cara/apps/calculator/__init__.py
Comment (pk-scs) [reply cara/apps/calculator/__init__.py:263]: No problem. Sure, I'll do it this way then and update the PR.

File: cara/apps/calculator/__init__.py
Comment (pelson) [inline cara/apps/calculator/__init__.py]: Would you mind addding this trailing comma back please?

File: cara/apps/calculator/__init__.py
Comment (pelson) [inline cara/apps/calculator/__init__.py:133]: Normally, for such an important implementation, I would expect a comment here to tell the reader why we disable XSRF cookies. (i.e. the fact that this is an API not a form handler, etc.). Nothing too laborious, but enough to allow readers to have the full context.

Also, I assume the test would fail if this method got removed (which is highly possible if people think that it isn't doing anything)

**Review by pelson — APPROVED** [review]
Looks great to me. I've added two additional tweaks, but there is nothing on my side from preventing this being merged :+1:. @pingud98 - how to proceed?

**Comment by pingud98** [comment]
Hi @pk-scs and @pelson, As for the next steps - later today (or tonight), I'll update the API branch on the CERN gitlab with the latest commits and copy this MR over for the gitlab/master.
In parallel I will reinstate the mirroring (currently off because of LFS, but that issue has been solved), and so the code will come back to github.

File: cara/apps/calculator/__init__.py
Comment (pk-scs) [reply cara/apps/calculator/__init__.py]: Sure, sorry for the accident.

File: cara/apps/calculator/__init__.py
Comment (pk-scs) [reply cara/apps/calculator/__init__.py:133]: Good idea with the comment explaining why we disable XSRF cookies. I added a comment.

**Comment by pk-scs** [comment]
Thanks for the review and the useful comments! I integrated the last two points mentioned by [pelson](https://github.com/pelson). I guess the PR could then be good to merge. I could clean up the Git history of this PR before merging by reorganizing the commits (now it's 10 commits but conceptionally, 3 would be enough). Do you usually reorganize commits or merge PRs as they are after being approved?

**Comment by pelson** [comment]
Sure, go ahead an tidy up the commits, I can re-approve subsequently. :+1:

**Comment by pk-scs** [comment]
It seems that still a _workflow approval_ from a _maintainer_ is required to merge this PR. Could one of the maintainers give such a _workflow approval_? Thank you very much.

**Comment by pingud98** [comment]
Hi Pascal,

Sorry - I've been a bit behind schedule. In fact I'll put this
upstream into our gitlab repo and we'll do the MR there. For now
Github is still a force-push mirror from gitlab, so any PR here would
get overwritten. How is the commit tidying you wanted to do?

Best regards,
James

**Comment by pk-scs** [comment]
Hi James, 

Thanks for your reply and no worries. I rearranged the commit history and pushed everything 2 days ago (11.04.2022). Instead of 10 commits the PR has now 3 commits. I rewrote the history of the respective branch, but I guess that shouldn't be a problem. For me the PR would be ready to be merged.

Bests, Pascal

**Comment by pingud98** [comment]
Super, I'll get the 3 new commits upstreamed now.

**Comment by pingud98** [comment]
Pull request upstreamed to [CERN gitlab repo](https://gitlab.cern.ch/cara/cara/-/merge_requests/347)

**Pull request closed without merging**
